etna/collections/models.py

```diff
@@ -309,3 +314,122 @@ def record(self):
         FieldPanel("teaser_image"),
         FieldPanel("description"),
     ]
+
+
+class PageTopic(Orderable):
+    """
+    This model allows any page type to be associated with one or more topics
+    in a way that retains the order of topics selected.
+
+    The ``sort_order`` field value from ``Orderable`` can be used to pull out
+    the 'first' topic to treat as the 'primary topic' for a page, and can also
+    used to prioritise items for a list of 'pages related to a topic'.
+
+    Just add `InlinePanel("page_topics")` to a page type's panel
+    configuration to use it!
+    """
+
+    page = ParentalKey(Page, on_delete=models.CASCADE, related_name="page_topics")
+    topic = models.ForeignKey(
+        TopicExplorerPage,
+        verbose_name=_("topic"),
+        related_name="topic_pages",
+        on_delete=models.SET_NULL,
+        null=True,
+    )
+
+
+class PageTimePeriod(Orderable):
+    """
+    This model allows any page type to be associated with one or more topics
+    in a way that retains the order of topics selected.
+
+    The ``sort_order`` field value from ``Orderable`` can be used to pull out
+    the 'first' topic to treat as the 'primary topic' for a page, and can also
+    used to prioritise items for a list of 'pages related to a topic'.
+
+    Just add `InlinePanel("page_time_periods")` to a page type's panel
+    configuration to use it!
+    """
+
+    page = ParentalKey(Page, on_delete=models.CASCADE, related_name="page_time_periods")
+    time_period = models.ForeignKey(
+        TimePeriodExplorerPage,
+        verbose_name=_("time period"),
+        related_name="time_period_pages",
+        on_delete=models.SET_NULL,
+        null=True,
+    )
+
+
+class TopicalPageMixin:
+    """
+    A mixin for pages that use the ``PageTopic`` and ``PageTimePeriod`` models
+    in order to be associated with one or many topics/time periods. It simply
+    adds a few properies to support robust, efficient access the related topic
+    and time period pages.
+    """
+
+    @classmethod
+    def get_time_periods_inlinepanel(
+        cls, max_num: Optional[int] = 4, min_num: Optional[int] = None
+    ) -> InlinePanel:
+        kwargs = {
+            "heading": _("Related time periods"),
+            "max_num": max_num,
+        }
+        if min_num:
+            kwargs["min_num"] = min_num
```

Comment:
Suggestion: we could reduce the time complexity by one parity check to just assign `min_num` from the signature (set to 0) and assigning that to the kwargs. `On` instead of `(On + 1)` operations in the worse case

Reply:
I didn't like the idea of having `0` as a default value because it's a bit meaningless really. However, you prompted me to check out [`InlinePanel.__init__()`](https://github.com/wagtail/wagtail/blob/main/wagtail/admin/panels/inline_panel.py#L21), and it looks like it will accept a `None` value just fine, so I can pass it straight along without changing the default value :)

Reply:
Nice!